src/Servers/Kestrel/samples/PlaintextApp/Startup.cs

```diff
@@ -16,18 +18,26 @@ public class Startup
 
         public void Configure(IApplicationBuilder app)
         {
-            app.Run((httpContext) =>
+            app.Run(async (httpContext) =>
             {
                 var response = httpContext.Response;
                 response.StatusCode = 200;
                 response.ContentType = "text/plain";
+                response.ContentLength = _helloWorldBytes.Length;
 
-                var helloWorld = _helloWorldBytes;
-                response.ContentLength = helloWorld.Length;
-                return response.Body.WriteAsync(helloWorld, 0, helloWorld.Length);
+                var pipe = response.BodyPipe;
+                Write(pipe, _helloWorldBytes);
+                await pipe.FlushAsync();
```

Comment:
FlushAsync returns a `ValueTask<T>` right? Could `ValueTask.IsCompletedSuccessfully` be tested so that the async state machine is only created when its false? Or does the nature of pipelines mean it always be false?

Reply:
Could do; technically you should then also call `GetAwaiter().GetResult()` on the ValueTask to communicate consumption back to the `IValueTaskSource` 

Not sure it matters for pipelines; but people might use it as reference.

Reply:
Why would a state machine be created for a completed task?

This feels like a slight over-optimization for a sample.

Reply:
Because it forces the method to be marked `async` which then always becomes a statemachine.

> This feels like a slight over-optimization for a sample.

Its the go to sample for people that want to test plaintext without working out what's going on in  aspnet/benchmarks; as that's 100 scenerios in one app; with lots of config switching and conditional includes.

Reply:
What about doing result.AsTask()?

Reply:
Is a result task, it would allocate a `Task<FlushResult>`. 

Thought about adding an extension method called `.AsNonGenericTask`; but that seemed a terrible name.

Reply:
I see. I would wait for other opinions, personally I still think sample should be simple.

Reply:
I still don't understand the desire fore `AsNonGenericTask()`. So what if `AsTask()` returns `Task<FlushResult>` instead of a `Task` when it is a `Task`? It's not like it has to allocate more to implicitly convert to a `Task`.

Are you just worried about the larger object size? Are you sure the local `AwaitResult` function really more efficient than `AsTask()`?

Reply:
Its a chunky ol'block of code https://github.com/dotnet/coreclr/blob/9aa68da7d3dcd24761d03c4f41720194ab3638b8/src/System.Private.CoreLib/shared/System/Threading/Tasks/ValueTask.cs#L522-L645 but it doesn't make it much nicer.

What I wanted to do with `AsNonGenericTask()` is remove the whole lot including `IsCompletedSucessfully`; so it would just be:

```csharp
return response.BodyPipe.WriteAsync(payload).AsNonGenericTask();
```
And that would handle the optimization to `Task.CompletedTask`

Reply:
Say what you will about GetTaskForValueTaskSource(), but at least it doesn't introduce an async state machine like the local AwaitResult() function, so it's still not obvious to me that AwaitResult() is more efficient than AsTask().

The desire for AsNonGenericTask makes a lot more sense to me now though. Not having to write the IsCompletedSucessfully logic ourselves would be nice. I'm still hoping that someday the awaited pattern we use all over the place will become a relic of the past because it's no longer necessary.

Reply:
> This feels like a slight over-optimization for a sample.

Yeah, for a sample the original code was fine. I was coming at this code from the perspective that it is going to be run in benchmarks. It will still be useful to measure if this is faster is the plaintext benchmark.